### ISSUE TITLE
Lots of CLOSE_WAITS to Vault

### ISSUE THREAD

**Issue by joaquin386** [petems/petems-hiera_vault#25]
Hi,

Version: 
```
Puppet: 6.1
Vault: 1.0.1
Gem Vault: 0.12.0
Hiera_vault: Latest repo version
```

Install the hiera_vault and everything working fine. BUT there is a huge number of CLOSE_WAITS. Seems the function doesn't close the connection once it finish. Noticing as well that after removing hiera.yaml all ESTABLISH move to CLOSE_WAIT but the connection still not close.

Before removing hiera.yaml (after removingthe difference is that all connections are CLOSE_WAIT)
10.100.10.101-> Puppetserver
10.100.2.157 -> Vault
```
netstat -an | grep 10.100.2.157
tcp        0      0 10.100.10.101:39110     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37442     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38504     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38530     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38758     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38000     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38506     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38008     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39212     10.100.2.157:8200       ESTABLISHED
tcp        0      0 10.100.10.101:39076     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37426     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38006     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37638     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39072     10.100.2.157:8200       ESTABLISHED
tcp        0      0 10.100.10.101:39236     10.100.2.157:8200       ESTABLISHED
tcp        0      0 10.100.10.101:39234     10.100.2.157:8200       ESTABLISHED
tcp        0      0 10.100.10.101:39250     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37888     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39078     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:38772     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39232     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:38502     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38792     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37440     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37886     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37094     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39238     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37822     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39240     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37940     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39230     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37928     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39074     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37116     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37460     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38764     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37648     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39242     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37270     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38498     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38766     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37932     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38624     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38642     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37448     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38500     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39070     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37452     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38018     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39224     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37620     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37884     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38528     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38524     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39246     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37130     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37936     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39098     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:38770     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38768     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37838     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38014     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:38756     10.100.2.157:8200       CLOSE_WAIT 
tcp        0      0 10.100.10.101:39208     10.100.2.157:8200       ESTABLISHED
tcp       54      0 10.100.10.101:37332     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37938     10.100.2.157:8200       CLOSE_WAIT 
tcp       54      0 10.100.10.101:37852     10.100.2.157:8200       CLOSE_WAIT 
```
And here hiera
```
---
version: 5
hierarchy:
  - name: "Hiera-vault lookup"
    lookup_key: hiera_vault
    options:
      confine_to_keys:
        - '^vault_.*'
        - '^.*_password$'
        - '^password.*'
      ssl_verify: true
      ssl_ca_cert: /etc/pki/tls/cert.pem
      address: https://vault.server.com:8200
      token: blablabla
      default_field: value
      mounts:
        secret:
          - puppet/%{::trusted.certname}
```

Citing the Red Hat docs:
```
CLOSE_WAIT - Indicates that the server has received the first FIN signal from the client and the connection is in the process of being closed. This means the socket is waiting for the application to execute close(). A socket can be in CLOSE_WAIT state indefinitely until the application closes it. Faulty scenarios would be like a file descriptor leak: server not executing close() on sockets leading to pile up of CLOSE_WAIT sockets.
```

**Comment by joaquin386** [comment]
just instll it will let it run for a while

**Comment by joaquin386** [comment]
@petems did you push the changes to master? (Cause the branch is not there anymore: readd_caching_vault_object)

**Comment by petems** [comment]
Can you completely stop puppetserver, then `kill -9` all the processes that already existed, then run again using that branch?

Just want to make sure that the existing processes aren't left over from before.

**Comment by petems** [comment]
If that doesn't work, do that again and kill all the processes, then we can try a new branch with an experimental change (`readd_caching_vault_object_klass`)

**Comment by joaquin386** [comment]
```
# netstat -anp | grep 10.100.2.15[7-9] | grep CLOSE | wc -l
136
# systemctl restart puppetserver
# netstat -anp | grep 10.100.2.15[7-9] | grep CLOSE | wc -l
0
```
Restarted puppetserver. Now CLOSE are 0. I have a three cluster vault node.

**Comment by joaquin386** [comment]
It is with the ```readd_caching_vault_object```. So far so good.
```
# netstat -anp | grep 10.100.2.15[7-9] | grep CLOSE | wc -l
11
# netstat -anp | grep 10.100.2.157 | grep CLOSE | wc -l
6
```
Will let it run for the whole weekend and come back on Monday.

**Comment by joaquin386** [comment]
I would say that ```readd_caching_vault_object``` is working fine.

**Comment by petems** [comment]
Hooray! I'll merge now then 😄

**Comment by joaquin386** [comment]
Thx for that @petems

**Comment by petems** [comment]
Will cut a release now 👍

**Comment by petems** [comment]
Released in 0.2.0: closing

**Comment by petems** [comment]
Fixed by #26

**Comment by StanvanHoorn** [comment]
This seems to have surfaced again; Can anyone have a look?
